crates/ra_db/src/input.rs

```diff
@@ -83,6 +83,17 @@ pub struct CrateGraph {
 #[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
 pub struct CrateId(pub u32);
 
+pub struct CrateName(SmolStr);
+
+impl<T: AsRef<str>> From<T> for CrateName {
+    fn from(name: T) -> Self {
+        // For root projects with dashes in their name,
+        // cargo metadata does not do any normalization
+        // so we do it ourselves
+        Self(SmolStr::new(name.as_ref().replace('-', "_")))
+    }
+}
```

Comment:
Hm, I don't like the fact that we just unconditionally normalize here...

Like, if one day `cargo metadata` starts using `-` in the `deps` array, we should notice this and report a bug to cargo, rather than do a silent conversion.

I propose the following API

```
impl CrateName {
    // We use this for deps and in fixtures, with just `.unwrap`ing the result
    fn new(name: &str) -> Result<CrateName>;
    // We use this in that single place where we genuinely expect a "malformed" crate name. 
    fn normalize_dashes(name: &str) -> CrateName;
}
```

Reply:
Feels rather odd to use just `unwrap` there, but here you go.

That also makes the new test I've added kind of obsolete, but I have doubts that I can add a new one in the `ra_project_model` `lib.rs`.
Any pointers?

Reply:
>Feels rather odd to use just unwrap everywhere, but here you go.

It should feel *right*. `.unwrap()` means "I am asserting an invariant here", it's like a more concise version of `assert!(!name.contains("-"))` on every call-site. 

>Any pointers?

I think it's fine not to test this. We make sure that this does not break by encapsulating logic into `CrateName`, types are our tests here.

Reply:
Great, thanks for the tips, useful and assuring, as always :)